feat: replace bb_cli_bench and op-count-time build with bb and runtime flag

johnathan79717 opened the pull request:
`bb` takes --print_benchmark or --benchmark_out=<json>. If either is given, it will run with op-count-time in a benchmark. `--print_benchmark` only prints the total time but not the op-count-time though, I'm not sure how to get that from Google benchmark.

Other binaries can be run with op-count-time with env `BB_USE_OP_COUNT_TIME=1`.

There's still a 1% overhead even when the env or the runtime flag is disabled because we need to check the flag anyway.

Fixes https://github.com/AztecProtocol/barretenberg/issues/1458

johnathan79717 commented:
Ran on the benchmark machine via `scripts/benchmark_remote.sh client_ivc_bench './client_ivc_bench --benchmark_filter=ClientIVCBench/Full/6' clang16 build`

This branch:

| Runtime | Peak Memory |
| --- | --- |
| 29,836 ms | 2852.46 MiB |
| 29,721 ms | 2836.17 MiB |
| 29,810 ms | 2852.46 MiB |

next:

| Runtime | Peak Memory |
| --- | --- |
| 29,015 ms | 2553.28 MiB |
| 29,113 ms | 2537.28 MiB |
| 29,076 ms | 2530.29 MiB |

There's about a 3% slowdown and 300MB memory overhead, which does not seem very low.
@ludamad maybe it's not worth collecting op-count-time in all builds?

johnathan79717 committed:
feat: collect op-count-time in all builds

johnathan79717 committed:
not profile

johnathan79717 committed:
split parse_and_run_cli_command

johnathan79717 committed:
run as benchmark

johnathan79717 committed:
reassemble args for benchmark

johnathan79717 committed:
mute console

johnathan79717 committed:
remove bb_cli_bench and op-count-time

johnathan79717 committed:
Update README

johnathan79717 committed:
remove no-avm and fix benchmark_example script

johnathan79717 commented:
Will take a look at test failure first.

johnathan79717 committed:
fix paths

johnathan79717 committed:
fix wasm builds

johnathan79717 commented:
ci-full passed, ready for review.

johnathan79717 commented:
I checked with @ludamad  offline, we also want to collect count by default but we can get rid of cycle. I'll update the PR accordingly.

johnathan79717 committed:
collect count by default but remove cycle

johnathan79717 commented:
@ludamad ci-full passed. Ready for review again.

ludamad commented on barretenberg/cpp/bootstrap.sh:
```diff
@@ -291,18 +290,18 @@ case "$cmd" in
     # Intended only for dev usage. For CI usage, we run yarn-project/end-to-end/bootstrap.sh bench.
     # Sample usage (CI=1 required for bench results to be visible; exclude NO_WASM=1 to run wasm benchmarks):
-    # CI=1 NO_WASM=1 NATIVE_PRESET=op-count-time ./barretenberg/cpp/bootstrap.sh bench_ivc transfer_0_recursions+sponsored_fpc
+    # CI=1 NO_WASM=1 ./barretenberg/cpp/bootstrap.sh bench_ivc transfer_0_recursions+sponsored_fpc
```
Do these instructions work?

ludamad commented on barretenberg/cpp/scripts/benchmark_example_ivc_flow_remote.sh:
```diff
@@ -1,13 +1,13 @@
 #FLOW=${2:-"ecdsar1+amm_add_liquidity_1_recursions+sponsored_fpc"}
 #FLOW=${2:-"ecdsar1+transfer_1_recursions+private_fpc"}
 #FLOW=${2:-"ecdsar1+transfer_1_recursions+sponsored_fpc"}
 #FLOW=${2:-"ecdsar1+transfer_1_recursions+sponsored_fpc"}
```
Is this gonna work? The benchmark format json requires a gtest env?

johnathan79717 replied:
Yes, it works. I'm not sure which gtest env you were referring to.

ludamad replied:
Sorry, google benchmark env - i saw later you had linked it

ludamad commented on barretenberg/cpp/src/barretenberg/bb/cli.cpp:
```diff
@@ -80,6 +84,100 @@ void print_subcommand_options(const CLI::App* sub)
+
+// This is used to suppress the default output of Google Benchmark.
+// This is useful to run this benchmark without altering the stdout result of our simulated bb cli command.
+// We instead use the `--benchmark_out` flag to output the results to a file.
```
Hmm how much does linking Google benchmark take in size? I was thinking we'd do something simpler.

johnathan79717 replied:
Not using Google benchmark anymore. We will not have the total time though.

ludamad commented:
Looks nice overall but the dependency feels overkill vs just creating a json string?

johnathan79717 commented:
Hmm I thought we had to go through Google Benchmark to get the stats, but maybe we don't. I can look into it.

ludamad commented:
We have a global table with all the stats - should be able to go through that. Ah if it's 1% for native, what's the slowdown for WASM? it could be significantly worse, in which case we should maybe instead (as it sounds like the conditional might be slow enough) remove any benchmark commands from stuff like the field class, elliptic curve class (point/affine_element) etc as they are too low level.

johnathan79717 commented:
> Ah if it's 1% for native, what's the slowdown for WASM?

I excluded WASM in the previous attempt but can experiment.

johnathan79717 committed:
revert cli.cpp

johnathan79717 committed:
print op count as json

ludamad commented:
Got it, no worries if wasm is disabled missed that

johnathan79717 committed:
write to file

johnathan79717 committed:
make compatible with analyze script

johnathan79717 committed:
revert to one print method but make analyze script compatible

johnathan79717 committed:
fix ci_benchmark_ivc_flows.sh

ludamad reviewed (approved):
LGTM thanks for this!

Commit:
feat: merge via append the ecc ops in hiding kernel (#16267)

Modified the logic in CIVC class and related tests to support merging
via append the subtable of ecc ops coming from the hiding kernel to the
total table. With this modification, all CIVC tests have to accumulate a
meaningful number of circuit to reach a tail kernel (which recursively
verifies a proof labelled as `PG_TAIL`). This is due to the fact that it
is the tail kernel that will now start with a no-op in order to ensure
that we can correctly shift the polynomials representing the op queue in
translator.

johnathan79717 committed:
fix wasm builds

johnathan79717 committed:
fix wasm benchmark script

Commit:
[empty] Start merge-train. Choo choo.

ludamad committed:
Revert "feat: merge via append the ecc ops in hiding kernel (#16267)"

This reverts commit ec8e27f8fc1742e98347140bb1fa0c095ceec043.
